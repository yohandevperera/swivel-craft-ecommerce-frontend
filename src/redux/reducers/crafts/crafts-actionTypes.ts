/**
 * Usage - This file will manipulate the redux global actions types.
 *
 * Description - This file consists for all the redux global actions types that will be
 * used to manipulate the redux states
 */

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  CRAFT_LOAD_START: "CRAFT_LOAD_START",
  CRAFT_LOAD_SUCCESS: "CRAFT_LOAD_SUCCESS",
  CRAFT_LOAD_ERROR: "CRAFT_LOAD_ERROR",
  SINGLE_CRAFT_LOAD_SUCCESS: "SINGLE_CRAFT_LOAD_SUCCESS",
};
