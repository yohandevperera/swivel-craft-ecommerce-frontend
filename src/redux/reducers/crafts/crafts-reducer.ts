import actionTypes from "./crafts-actionTypes";
import initialState from "./crafts-intialStates";

/**
 * Usage - This file will manipulate the defined global redux states.
 *
 * Description - All the defined global redux states can be manipulated using this file
 */

/**
 *
 * Description - The function that will be used as the state reducer to manipulate
 * the defined states
 *
 * @param state @typedef object
 * @param action @typedef any
 */
const craftReducer = (state = initialState, action: any) => {
  const { type, payload } = action;
  switch (type) {
    case actionTypes.CRAFT_LOAD_START:
      return {
        ...state,
        isLoading: true,
        crafts: null,
        errorMessage: null,
      };
    case actionTypes.CRAFT_LOAD_SUCCESS:
      return {
        ...state,
        isLoading: false,
        crafts: payload,
        craftList: payload,
      };
    case actionTypes.SINGLE_CRAFT_LOAD_SUCCESS:
      return {
        ...state,
        isLoading: false,
        craft: payload,
      };
    case actionTypes.CRAFT_LOAD_ERROR:
      return {
        ...state,
        isLoading: false,
        errorMessage: payload,
      };

    default:
      return state;
  }
};

export default craftReducer;
