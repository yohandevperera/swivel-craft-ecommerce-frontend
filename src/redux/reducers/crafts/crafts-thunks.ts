import { Dispatch } from "redux";
import actions from "./crafts-actions";
import {
  CraftType,
  getAllCrafts,
  createCraft,
  editCraft,
  deleteCraft,
  getCraft,
} from "../../../services/crafts";
import _ from "lodash";

/**
 * Usage - This file will be used to communcate with the API services via redux.
 *
 * Description - This file will act as a mediator for the services and the redux global states
 */

/**
 * Description and Usage - This function that will be used load all the crafts
 * to the redux state
 *
 * @param dispatch @typedef Dispatch
 */
export const loadAllCrafts = () => (dispatch: Dispatch) => {
  dispatch(actions.craftLoadStart());
  getAllCrafts()
    .then((response) => {
      if (_.isEmpty(response) || _.isNull(response)) {
        throw Error("load all crafts response null");
      }
      const crafts = response.data.data;
      dispatch(actions.craftLoadSuccess(crafts));
    })
    .catch((error) => {
      dispatch(actions.craftLoadError(error.message));
      throw Error(error.message);
    });
};

/**
 * Description and Usage - This function that will be used create crafts
 * by fetching the payload from the redux state
 *
 * @param dispatch @typedef Dispatch
 * @param crafts @typedef Omit<CraftType, "photo" | "_id">
 */
export const createCrafts =
  (craft: Omit<CraftType, "photo" | "_id">) => (dispatch: Dispatch) => {
    dispatch(actions.craftLoadStart());
    createCraft(craft)
      .then((response) => {
        if (_.isEmpty(response) || _.isNull(response)) {
          throw Error("create craft response null");
        }
        const createdResponse = response.data;
        dispatch(actions.singlecraftLoadSuccess(createdResponse));
      })
      .catch((error) => {
        dispatch(actions.craftLoadError(error.message));
        throw Error(error.message);
      });
  };

/**
 * Description and Usage - This function can be used edit crafts
 * by fetching the payload from the redux state
 *
 * @param dispatch @typedef Dispatch
 * @param craft @typedef Omit<CraftType, "photo" | "_id">
 * @param craftId @typedef string
 */
export const editCrafts =
  (craft: Omit<CraftType, "photo" | "_id">, craftId: string) =>
  (dispatch: Dispatch) => {
    dispatch(actions.craftLoadStart());
    editCraft(craft, craftId)
      .then((response) => {
        if (_.isEmpty(response) || _.isNull(response)) {
          throw Error("edit craft response null");
        }
        const createdResponse = response.data;
        dispatch(actions.singlecraftLoadSuccess(createdResponse));
      })
      .catch((error) => {
        dispatch(actions.craftLoadError(error.message));
        throw Error(error.message);
      });
  };

/**
 * Description and Usage - This function that will be used remove crafts
 * by fetching the craft id as the payload from the redux state
 *
 * @param dispatch @typedef Dispatch
 * @param id @typedef string
 */
export const removeCraft = (id: string) => (dispatch: Dispatch) => {
  dispatch(actions.craftLoadStart());
  deleteCraft(id)
    .then((response) => {
      if (_.isEmpty(response) || _.isNull(response)) {
        throw Error("remove craft response null");
      }
      const deletedResponse = response.data;
      dispatch(actions.singlecraftLoadSuccess(deletedResponse));
    })
    .catch((error) => {
      dispatch(actions.craftLoadError(error.message));
      throw Error(error.message);
    });
};

/**
 * Description and Usage - This function that will be used to fetch
 * a single craft for a given craft id
 *
 * @param dispatch @typedef Dispatch
 * @param id @typedef string
 */
export const getSingleCraft = (id?: string) => (dispatch: Dispatch) => {
  dispatch(actions.craftLoadStart());
  getCraft(id)
    .then((response) => {
      if (_.isEmpty(response) || _.isNull(response)) {
        throw Error("get single craft response null");
      }
      const craft = response.data.data;
      dispatch(actions.singlecraftLoadSuccess(craft));
    })
    .catch((error) => {
      dispatch(actions.craftLoadError(error.message));
      throw Error(error.message);
    });
};
