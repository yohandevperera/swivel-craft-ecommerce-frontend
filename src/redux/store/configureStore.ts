import {
  applyMiddleware,
  legacy_createStore as createStore,
  compose,
} from "redux";
import thunk from "redux-thunk";
import rootReducer from "../reducers/rootReducer";

/**
 * Description and Usage - This file will include all the redux
 * store configurations
 *
 */

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

const configureStore = () => {
  const middlewares = [thunk];
  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  const enhancers = composeEnhancers(applyMiddleware(...middlewares));
  const store = createStore(rootReducer(), enhancers);

  return store;
};

export default configureStore;
