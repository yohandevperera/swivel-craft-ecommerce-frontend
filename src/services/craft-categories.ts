import { invoke } from "../utils/api-adapter";

/**
 *
 * Usage - This file will be used to communicate to the backend API endpoints
 *
 * Description - This file consists of all the API service call functions for the CraftCategorysCategories resource
 *
 */

/**
 * Description - A global type thats used as
 * CraftCategory add payload and to fetch CraftCategorysCategories as well
 */
export interface CraftCategoryType {
  _id: string;
  name: string;
}

/**
 * Description and Usage - This function will be used
 * to fetch all the CraftCategorysCategories from the backend API endpoint
 */
export async function getAllCraftCategories() {
  return invoke("api/craft-categories/", "get", {});
}

/**
 * Description and Usage - This function will be used
 * to create an CraftCategory using the backend API endpoint
 *
 * @param CraftCategory @typedef Omit<CraftCategoryType, "_id" | "photo">
 */
export async function createCraftCategory(
  craftCategory: Omit<CraftCategoryType, "_id">
) {
  return invoke("api/craft-categories/", "post", craftCategory);
}

/**
 * Description and Usage - This function will be used
 * to edit an CraftCategory using the backend API endpoint
 *
 * @param CraftCategory @typedef Omit<CraftCategoryType, "_id" | "photo">
 * @param id @typedef string
 */
export async function editCraftCategory(
  craftCategory: Omit<CraftCategoryType, "_id" | "photo">,
  id?: string
) {
  return invoke(`api/craft-categories${id}`, "patch", craftCategory);
}

/**
 * Description and Usage - This function will be used
 * to remove an CraftCategory using the backend API endpoint
 *
 * @param id @typedef string
 */
export async function deleteCraftCategory(id: string) {
  return invoke(`api/craft-categories/${id}`, "delete", {});
}

/**
 * Description and Usage - This function will be used
 * to fetch a single CraftCategory for a given id using the backend API endpoint
 *
 * @param id @typedef string
 */
export async function getCraftCategory(id?: string) {
  return invoke(`api/craft-categories/get-craft-category/${id}`, "get", {});
}

export async function searchCraftCategoryByName(name?: string) {
  return invoke(`api/craft-categories/find-by-name/${name}`, "get", {});
}

export async function getCraftCategoriesByName() {
  return invoke(`api/craft-categories/get-category-by-name`, "get", {});
}
