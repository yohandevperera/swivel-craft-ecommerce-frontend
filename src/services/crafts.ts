import { invoke } from "../utils/api-adapter";

/**
 *
 * Usage - This file will be used to communicate to the backend API endpoints
 *
 * Description - This file consists of all the API service call functions for the crafts resource
 *
 */

/**
 * Description - A global type thats used as
 * craft add payload and to fetch crafts as well
 */
export interface CraftType {
  _id: string;
  name: string;
  categoryId: string;
  description: string;
  qty: number;
  price: number;
  photo: any;
}

/**
 * Description and Usage - This function will be used
 * to fetch all the crafts from the backend API endpoint
 */
export async function getAllCrafts() {
  return invoke("api/crafts/", "get", {});
}

/**
 * Description and Usage - This function will be used
 * to create an craft using the backend API endpoint
 *
 * @param craft @typedef Omit<CraftType, "_id" | "photo">
 */
export async function createCraft(craft: Omit<CraftType, "_id" | "photo">) {
  return invoke("api/crafts/", "post", craft);
}

/**
 * Description and Usage - This function will be used
 * to edit an craft using the backend API endpoint
 *
 * @param craft @typedef Omit<CraftType, "_id" | "photo">
 * @param id @typedef string
 */
export async function editCraft(
  craft: Omit<CraftType, "_id" | "photo">,
  id?: string
) {
  return invoke(`api/crafts/${id}`, "patch", craft);
}

/**
 * Description and Usage - This function will be used
 * to remove an craft using the backend API endpoint
 *
 * @param id @typedef string
 */
export async function deleteCraft(id: string) {
  return invoke(`api/crafts/${id}`, "delete", {});
}

/**
 * Description and Usage - This function will be used
 * to fetch a single craft for a given id using the backend API endpoint
 *
 * @param id @typedef string
 */
export async function getCraft(id?: string) {
  return invoke(`api/crafts/${id}`, "get", {});
}
