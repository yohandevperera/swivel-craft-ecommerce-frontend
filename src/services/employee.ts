import { invoke } from "../utils/api-adapter";

/**
 *
 * Usage - This file will be used to communicate to the backend API endpoints
 *
 * Description - This file consists of all the API service call functions
 *
 */

/**
 * Description - A global type thats used as
 * employee add payload and to fetch employees as well
 */
export interface EmployeeType {
  _id: string;
  photo: any;
  firstname: string;
  lastname: string;
  email: string;
  phone: string;
  gender: "M" | "F" | "Male" | "Female" | string;
}

/**
 * Description and Usage - This function will be used
 * to fetch all the employees from the backend API endpoint
 */
export async function getAllEmployees() {
  return invoke("api/employees/", "get", {});
}

/**
 * Description and Usage - This function will be used
 * to create an employee using the backend API endpoint
 *
 * @param employee @typedef Omit<EmployeeType, "_id" | "photo">
 */
export async function createEmployee(
  employee: Omit<EmployeeType, "_id" | "photo">
) {
  return invoke("api/employees/", "post", employee);
}

/**
 * Description and Usage - This function will be used
 * to edit an employee using the backend API endpoint
 *
 * @param employee @typedef Omit<EmployeeType, "_id" | "photo">
 * @param id @typedef string
 */
export async function editEmployee(
  employee: Omit<EmployeeType, "_id" | "photo">,
  id?: string
) {
  return invoke(`api/employees/${id}`, "put", employee);
}

/**
 * Description and Usage - This function will be used
 * to remove an employee using the backend API endpoint
 *
 * @param id @typedef string
 */
export async function deleteEmployee(id: string) {
  return invoke(`api/employees/${id}`, "delete", {});
}

/**
 * Description and Usage - This function will be used
 * to fetch a single employee for a given id using the backend API endpoint
 *
 * @param id @typedef string
 */
export async function getEmployee(id?: string) {
  return invoke(`api/employees/${id}`, "get", {});
}
