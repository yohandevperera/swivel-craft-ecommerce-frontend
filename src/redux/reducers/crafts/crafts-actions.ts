import { CraftType } from "../../../services/crafts";
import actionTypes from "./crafts-actionTypes";

/**
 * Usage - This file will manipulate the redux global actions.
 *
 * Description - This file consists for all the redux global actions that will be used
 * to manipulate data
 */

/**
 * function that will handle the loading action
 */
const craftLoadStart = () => ({
  type: actionTypes.CRAFT_LOAD_START,
});

/**
 * function that will handle the load success action
 */
const craftLoadSuccess = (crafts: CraftType[]) => ({
  type: actionTypes.CRAFT_LOAD_SUCCESS,
  payload: crafts,
});

/**
 * function that will handle the load error action
 */
const craftLoadError = (errorMessage: any) => ({
  type: actionTypes.CRAFT_LOAD_ERROR,
  payload: errorMessage,
});

/**
 * function that will handle single craft load success action
 */
const singlecraftLoadSuccess = (craft: CraftType) => ({
  type: actionTypes.SINGLE_CRAFT_LOAD_SUCCESS,
  payload: craft,
});

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  craftLoadStart,
  craftLoadSuccess,
  craftLoadError,
  singlecraftLoadSuccess,
};
