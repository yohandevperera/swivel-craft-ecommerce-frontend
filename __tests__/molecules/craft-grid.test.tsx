import "@testing-library/jest-dom";
import React from "react";
import AppRoutes from "../../src/components/organisms/app-routes";
import { CraftType } from "../../src/services/crafts";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import configureStore from "../../src/redux/store/configureStore";
import GridItem from "../../src/components/molecules/grid-item";
import renderer from "react-test-renderer";

// Sample data set to check the grid view
const craft: CraftType = {
  _id: "6435a142030702dfc3032abe",
  name: "Painted Wine Bottle Vases",
  categoryId: "64229dfa1f396b074aea09c4",
  description: "",
  qty: 10,
  price: 3000,
  photo: "",
};

// the store instance created to access the redux store
const store = configureStore();

// test snapshot to validate the data grid view
describe("Craft grid snapshot", () => {
  it("should render the craft gird with the meta test data", () => {
    const domTree = renderer
      .create(
        <>
          <Provider store={store}>
            <BrowserRouter>
              <AppRoutes />
              <GridItem item={craft} handleAddToCart={() => {}} />
            </BrowserRouter>
          </Provider>
        </>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
