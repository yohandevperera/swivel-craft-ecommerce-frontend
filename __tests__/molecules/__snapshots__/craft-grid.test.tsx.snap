// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Craft grid snapshot should render the craft gird with the meta test data 1`] = `
[
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1bwk6yd-MuiPaper-root-MuiCard-root"
  >
    <div
      className="css-7s17j4"
    >
      Craft Manager
    </div>
  </div>,
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-oonc48-MuiPaper-root-MuiCard-root"
    >
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style={
          {
            "height": 40,
            "marginLeft": "50%",
            "marginTop": "10%",
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={{}}
          />
        </svg>
      </span>
    </div>
  </div>,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-cz2w44-MuiPaper-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          className="MuiBox-root css-5u3pr2"
        >
          <img
            alt="complex"
            className="css-y9s1zk"
            src=""
          />
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-true css-114x5zz-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-grid-xs-true css-hfwr93-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
          >
            <div
              className="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-1wniyei-MuiTypography-root"
            >
              Painted Wine Bottle Vases
            </div>
            <p
              className="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-ejgmvm-MuiTypography-root"
            >
              64229dfa1f396b074aea09c4
            </p>
            <p
              className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              3000
            </p>
            <p
              className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              Items available  10
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              className="MuiStack-root css-vr01dl-MuiStack-root"
            >
              <div
                onClick={[Function]}
                style={{}}
              >
                <button
                  aria-label={null}
                  aria-labelledby={null}
                  className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element={true}
                  disabled={true}
                  onBlur={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={-1}
                  type="button"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="AddShoppingCartIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
